### PR TITLE
Integrate PrecomputedText API, add Performance Sample.

### LINKED ISSUE #8: PrecomputedText
Android P introduces new api to pre-compute text layout metrics, so rendering can be done faster.

Since Domic is intended to be used off-main thread this seems like a good api to integrate with.

Docs: https://developer.android.com/reference/android/text/PrecomputedText
ADB Podcast episode where they discuss this api: https://androidbackstage.blogspot.com/2018/05/episode-90-spanspanspanspan.html

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
Closes #8.

***Disclaimer***: Performance is not the reason of Domic's existence, however because of Domic's design we're able to achieve ridiculous things and I honestly believe that Domic will be 120 fps ready once we hit v1 (but don't quote me on that lol).

[PrecomputedText is an API](https://developer.android.com/reference/android/text/PrecomputedText) designed to allow pre-computing text layouting data before actual rendering on bg threads thus avoiding these computations on Main and/or rendering thread, which PERFECTLY fits into Domic's design

Note that there is a support version of this API https://developer.android.com/reference/androidx/core/text/PrecomputedTextCompat, but I'm not sure about the support lib relations with Domic yet, so for now we depend only on Framework APIs.

---

PR also adds ridiculous Performance Sample that renders 10 TextView counters and toggles 3 buttons at **insane rates** (each 0.999 ms, 1ms, 2ms, etc).

compressed gif

Systrace of this UI without Domic (same threading, only actual rendering switches to Main Thread):

(code in `samples.performance.regular` package)

screen shot 2018-07-09 at 1 09 36 am

Systrace of this UI with Domic and PrecomputedText:
(code in `samples.performance.domic` package)

screen shot 2018-07-09 at 1 10 59 am

Systrace of this UI with Domic and without PrecomputedText:

screen shot 2018-07-09 at 1 29 31 am

Note that **in this particular example Domic's diffing engine doesn't optimize anything** because each new rendered value is different from the old one, I want to showcase that in the sample some time in the future. 

**The sole reason for such good results is our optimized double-buffer rendering + sync with Chroreographer.**

There is also noticeable difference in gc kick-in frequency between Precomputed vs no Precomputed text systraces, I'll be taking a look later, might have been just runtime decision of ART.

Environment:

- Android Emulator 27.1.12.0 (Nexus 5X, API 28, Google Play Image)
- 3 GHz Intel Xeon W
- 64 GB 2666 MHz DDR4
- Radeon Pro Vega 64 16368 MB
- macOS 10.13.5

### CODE REVIEW

File: samples/performance/proguard-rules.pro [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Add project specific ProGuard rules here.
```
Comment (vanniktech) [inline samples/performance/proguard-rules.pro]: Empty that file?

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/AndroidDomicView.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import com.lyft.domic.android.AndroidButton
+import com.lyft.domic.android.AndroidCompoundButton
+import com.lyft.domic.android.AndroidEditText
+import com.lyft.domic.android.AndroidTextView
+import com.lyft.domic.api.rendering.Renderer
+import com.lyft.domic.samples.performance.R
+
+class AndroidDomicView(root: ViewGroup, renderer: Renderer) : DomicView {
+
+    override val counter0 by lazy { AndroidTextView(root.findViewById(R.id.counter0), renderer) }
```
Comment (vanniktech) [inline samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/AndroidDomicView.kt]: Do we need double locking here? Or would not `lazy(NONE)` suffice?

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/DomicActivity.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_performance)
+
+        val view = AndroidDomicView(findViewById(android.R.id.content), PerformanceApplication.renderer)
+
+        disposable += Single
+                .fromCallable { DomicViewModel(view) }
+                .subscribeOn(Schedulers.computation())
+                .subscribe { viewModel -> disposable += viewModel }
```
Comment (vanniktech) [inline samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/DomicActivity.kt]: `.subscribe { disposable += it }`?

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/DomicViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                .subscribe(view.checkBox.change::checked)
+
+        disposable += Observable
+                .interval(4, MILLISECONDS)
+                .map { it % 2L == 0L }
+                .subscribe(view.radioButton.change::checked)
+    }
+
+
+    override fun isDisposed() = disposable.isDisposed
```
Comment (vanniktech) [inline samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/DomicViewModel.kt]: Composition won't work with private members that aren't passed as a parameter in a constructor, right?

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/main/AndroidMainView.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import com.lyft.domic.android.AndroidButton
+import com.lyft.domic.api.rendering.Renderer
+import com.lyft.domic.samples.performance.R
+import com.lyft.domic.samples.performance.domic.DomicActivity
+import com.lyft.domic.samples.performance.regular.RegularActivity
+import io.reactivex.Completable
+import io.reactivex.android.schedulers.AndroidSchedulers.mainThread
+
+class AndroidMainView(private val root: ViewGroup, renderer: Renderer) : MainView {
+    override val regularUiButton by lazy { AndroidButton(root.findViewById(R.id.button_regular_ui), renderer) }
```
Comment (vanniktech) [inline samples/performance/src/main/java/com/lyft/domic/samples/performance/main/AndroidMainView.kt]: same as above

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/main/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+                .subscribe { viewModel -> disposable += viewModel }
+    }
+
+    override fun onDestroy() {
+        disposable.dispose()
+        super.onDestroy()
+    }
+
+}
+
```
Comment (vanniktech) [inline samples/performance/src/main/java/com/lyft/domic/samples/performance/main/MainActivity.kt]: Spacing police 🚓

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/main/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        val view = AndroidMainView(findViewById(android.R.id.content), PerformanceApplication.renderer)
+
+        disposable += Single
+                .fromCallable { MainViewModel(view) }
+                .subscribeOn(Schedulers.computation())
+                .subscribe { viewModel -> disposable += viewModel }
```
Comment (vanniktech) [inline samples/performance/src/main/java/com/lyft/domic/samples/performance/main/MainActivity.kt]: same

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/AndroidDomicView.kt (hunk 2 above)
Comment (artem-zinnatullin) [reply samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/AndroidDomicView.kt]: Absolutely not, I actually just need `counter0 = AndroidTextView(…)` here, will fix

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/DomicActivity.kt (hunk 3 above)
Comment (artem-zinnatullin) [reply samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/DomicActivity.kt]: Yeah, I'm just afraid others won't get what's getting disposed :D

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/DomicViewModel.kt (hunk 4 above)
Comment (artem-zinnatullin) [reply samples/performance/src/main/java/com/lyft/domic/samples/performance/domic/DomicViewModel.kt]: yeah :(

File: samples/performance/src/main/java/com/lyft/domic/samples/performance/main/MainActivity.kt (hunk 6 above)
Comment (artem-zinnatullin) [reply samples/performance/src/main/java/com/lyft/domic/samples/performance/main/MainActivity.kt]: Woop-woop!
It's the sound of da police!
Woop-woop!
That's the sound of the beast!

**Comment by artem-zinnatullin** [comment]
>Isn't it possible to reuse some of the Views / ViewModel between the comparision (normal + regular)?

It is possible to a degree, I could also have `List<TextView>` instead of separate TextViews but after all that will complicate code and debugging/tracing quite significantly and last thing I want in Performance Sample is complicated code

File: samples/performance/proguard-rules.pro (hunk 1 above)
Comment (artem-zinnatullin) [reply samples/performance/proguard-rules.pro]: But what if in future we'll have something there! (okok)

**Pull request merged**